src/dmd/dsymbolsem.d

```diff
@@ -4780,16 +4803,9 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         // Reflect this.type to f because it could be changed by findVtblIndex
         f = funcdecl.type.toTypeFunction();
 
-        /* Contracts can only appear without a body when they are virtual interface functions or abstract
-         */
-        if (!funcdecl.fbody && !funcdecl.isAbstract() &&
-            (funcdecl.fensure || funcdecl.frequire) &&
-            !(id && funcdecl.isVirtual()))
-        {
-            auto cd = parent.isClassDeclaration();
-            if (!(cd && cd.isAbstract()))
-                funcdecl.error("in and out contracts on non-virtual/non-abstract functions require function body");
-        }
+    Ldone:
+        if (!funcdecl.fbody && !funcdecl.allowsContractWithoutBody())
+            funcdecl.error("in and out contracts on non-virtual/non-abstract functions require function body");
```

Comment:
```
Test failed: 
expected:
----
fail_compilation/fail12901.d(11): Error: constructor fail12901.S.this in and out contracts require function body
----
actual:
----
fail_compilation/fail12901.d(11): Error: constructor fail12901.S.this in and out contracts on non-virtual/non-abstract functions require function body
----
```

How about:

```d
if (parent.isStructDeclaration()))
    funcdecl.error("in and out contracts require function body");
else
   funcdecl.error(""in and out contracts on non-virtual/non-abstract functions require function body");
```

Reply:
The negation in the error message can be rewritten as a positive:
```
Contracts can  only appear without a body when they are virtual interface functions or abstract.
```
:-)